src/client/QXmppMixManager.cpp

```diff
@@ -0,0 +1,1372 @@
+        // Check whether MIX is supported.
+        if (iq.features().contains(ns_mix_pam)) {
+            setSupportedByServer(true);
+
+            // Check whether MIX archiving is supported.
+            if (iq.features().contains(ns_mix_pam_archiving)) {
+                setArchivingSupportedByServer(true);
+            }
+        } else {
+            setSupportedByServer(false);
+            setArchivingSupportedByServer(false);
+        }
+    }
+
+    const auto jid = iq.from().isEmpty() ? client()->configuration().domain() : iq.from();
+
+    // Search for a MIX service and check what it supports.
+    // if none can be found, remove them from the cache.
+    if (!iq.features().contains(ns_mix)) {
+        removeService(jid);
+        return;
+    }
+
+    const auto identities = iq.identities();
+
+    for (const QXmppDiscoveryIq::Identity &identity : identities) {
+        // ' || identity.type() == "text"' is a workaround for older ejabberd versions.
+        if (identity.category() == u"conference" && (identity.type() == MIX_SERVICE_DISCOVERY_NODE || identity.type() == u"text")) {
+            Service service;
+            service.jid = iq.from().isEmpty() ? client()->configuration().domain() : iq.from();
+            service.channelsSearchable = iq.features().contains(ns_mix_searchable);
+            service.channelCreationAllowed = iq.features().contains(ns_mix_create_channel);
+
+            addService(service);
+            return;
+        }
+    }
+
+    removeService(jid);
+}
+
+///
+/// Sets whether the own server supports MIX.
+///
+/// \param supportedByServer whether MIX is supported by the own server
+///
+void QXmppMixManager::setSupportedByServer(bool supportedByServer)
+{
+    if (d->supportedByServer != supportedByServer) {
+        d->supportedByServer = supportedByServer;
+        Q_EMIT supportedByServerChanged();
+    }
+}
+
+///
+/// Sets whether the own server supports archiving messages via
+/// \xep{0313, Message Archive Management} of MIX channels the user participates in.
+///
+/// \param archivingSupportedByServer whether MIX messages are archived by the own server
+///
+void QXmppMixManager::setArchivingSupportedByServer(bool archivingSupportedByServer)
+{
+    if (d->archivingSupportedByServer != archivingSupportedByServer) {
+        d->archivingSupportedByServer = archivingSupportedByServer;
+        Q_EMIT archivingSupportedByServerChanged();
+    }
+}
+
+///
+/// Adds a MIX service.
+///
+/// \param service MIX service
+///
+void QXmppMixManager::addService(const Service &service)
+{
+    auto itr = std::find_if(d->services.begin(), d->services.end(), [&jid = service.jid](const Service &service) {
+        return service.jid == jid;
+    });
+
+    if (itr == d->services.end()) {
+        d->services.append(service);
+    } else if (*itr == service) {
+        return;
+    } else {
+        *itr = service;
+    }
+
+    Q_EMIT servicesChanged();
```

Comment:
```suggestion
    if (itr != d->services.end()) {
        if (*itr == service) {
            // do not emit signal
            return;
        }
        *itr = service;
    } else {
        d->services.append(service);
    }

    Q_EMIT servicesChanged();
```
might be nitpicking, but this would confuse my brain less